### ISSUE TITLE
timezones are not handled in DefaultBinding

### ISSUE THREAD

**Issue by BrentDouglas** [jOOQ/jOOQ#5504]
When setting and getting timestamps, times and dates from the result set `DefaultBinding` calls:
```
java.sql.Timestamp getTimestamp(int columnIndex);
java.sql.Time getTime(int columnIndex);
java.sql.Date getDate(int columnIndex);
void setTimestamp(int parameterIndex, java.sql.Timestamp x) throws SQLException;
void setTime(int parameterIndex, java.sql.Time x) throws SQLException;
void setDate(int parameterIndex, java.sql.Date x) throws SQLException;
```
rather than
```
java.sql.Timestamp getTimestamp(int columnIndex, Calendar cal);
java.sql.Time getTime(int columnIndex, Calendar cal);
java.sql.Date getDate(int columnIndex, Calendar cal);
void setTimestamp(int parameterIndex, java.sql.Timestamp x, Calendar cal) throws SQLException;
void setTime(int parameterIndex, java.sql.Time x, Calendar cal) throws SQLException;
void setDate(int parameterIndex, java.sql.Date x, Calendar cal) throws SQLException;
```
As a result the timezone offset is not accounted for and the retrieved value is incorrect.
This might be a PG only issue.

jooq pro 3.8.3
postgresql 9.5.5
oracle java 1.8.0_92
OSX

**Comment by BrentDouglas** [comment]
Nope, never mind, there was some wrapper code in our project I had missed that was causing the issue.

**Comment by BrentDouglas** [comment]
I'm going to reopen this, after a lot of being confused this was actually the cause of my issue. Implementing a binding that provides a correct calendar results in the correct values. I think that it is probably specific to PG.

**Comment by BrentDouglas** [comment]
I'll also note that in the bindings above the DateTime one works here as it writes the joda DateTime to the DB rather than the jdk Timestamp. The DefaultBinding would need to manually append the timezone info from the Calendar to the Timestamp when writing to a `timestamptz` column as it does not contain the timezone. This is a slightly different issue that only affects timestamptz columns and results in not having the timezone set, while the calendar part above affects both DateTime and LocalDates and gives them the wrong offset even when the timezone is set.

**Comment by chwassme** [comment]
It seems that we encountered the same problem in an Oracle DB.
jooq pro 3.8.1
oracle db 12c
oracle java 1.8.0_60
OSX

**Comment by chwassme** [comment]
I must admit that we have already an custom DateTimeBinding for "TIMESTAMP WITH TIMEZONE" in place, but it's not working correctly...

**Comment by lukaseder** [comment]
Indeed, the `TIMESTAMP WITH TIMEZONE` data type support is not yet completely implemented in jOOQ. Experimental `java.time` support has been added to jOOQ 3.7 (https://github.com/jOOQ/jOOQ/issues/4338), but we're not there yet to say we officially support it in all databases. In particular, the semantics of this data type may differ between databases, e.g. in PostgreSQL, it has `java.time.Instant` semantics, whereas in Oracle, it's a mixture between `java.time.OffsetDateTime` and `java.time.ZonedDateTime`.

I'm hoping this will be part of jOOQ 3.9

**Comment by BrentDouglas** [comment]
@chwassme i thought the oracle driver uses the stream api which doesn't accept a tz. I think you would have to manually fix the offset in the binding.

**Comment by BrentDouglas** [comment]
The more I think about this, I think that jooqs timestamp with timezone datatype should target either String, OffsetDateTime or some jooq specific type wrapping both the Timestamp and Timezone. The way it is at the moment there's no way to recover the lost information. You can still get the correct time by setting the session tz (in pg anyway) or converting to utc or something in the db before sending it but imo that's not a great solution. It would be better if it 'just worked' out of the box

**Comment by eatgrass** [comment]
now I am using 3.11.11, code generated with `javaTimeTypes=true` got a `LocalDateTime`, but the timestamp we access from database is Instant I think.

**Comment by lukaseder** [comment]
> @eatgrass  now I am using 3.11.11, code generated with javaTimeTypes=true got a LocalDateTime, but the timestamp we access from database is Instant I think.

Thank you for your comment. I'm not sure this is related. This issue here is not about code generation. Would you mind creating a new issue with some more details (including the DDL of the table, the full code generation configuration)?
